### ISSUE TITLE
Drop JDK7 Support

### ISSUE THREAD

**Issue by Apreld01** [immutables/immutables#1110]
Currently, the pom file looks like it's targeting JDK7, are there plans to upgrade the targeted JDK? It seems like dropping JDK7 support would be a prudent choice given the fact that premier support has been removed as of July 2019.

**Comment by bmhm** [comment]
Agree, there is not even extended support available from Oracle or IBM.

**Comment by asereda-gs** [comment]
As far as generated sources go (immutables not criteria/mongo repositories) we still need to support java 7 because of Android platform.  That doesn't mean processor can't be java 8. 

Also you can still generate java7 and java8 immutables classes out of box. 

Is there any particular reason you want to drop JDK7 support (I presume for processor) ?

**Comment by Apreld01** [comment]
Android Studio has adopted jdk 8 for a few years now. Is this consideration still realistic? Would like to remove older dependencies( java7) from project and use java lambda features.

**Comment by elucash** [comment]
is there any specific dependencies preventing from using java 8 or its features, or is it merely a purge of any pre-8 bytecode in jars? please, share the scenario were you're experience locking-in pre-8 stuff. For example, on java 8 we generate quite adequate code, can use java.util.Optional etc, and as API we have only annotation jar (no runtime required, frankly we don't have one for core Immutables, but can use Guava if available etc), annotations are agnostic to any java7/java8 differences.

**Comment by Apreld01** [comment]
JDK7 support has already died out, a lot of open source projects have moved onto 1.9+. From a maintenance point, adopting a higher version would also allow a purge of older dependencies.  Happy to take a crack at it myself in a PR if there is support from core maintainers.

**Comment by asereda-gs** [comment]
According to [Android distribution dashboard](https://developer.android.com/about/dashboards). This is current (as of Nov'19) Android OS distribution usage: 

| Codename           | API | Distribution |
|--------------------|-----|--------------|
| Gingerbread        | 10  | 0.30%        |
| Ice Cream Sandwich | 15  | 0.30%        |
| Jelly Bean         | 16  | 1.20%        |
|                    | 17  | 1.50%        |
|                    | 18  | 0.50%        |
| KitKat             | 19  | 6.90%        |
| Lollipop           | 21  | 3.00%        |
|                    | 22  | 11.50%       |
| Marshmallow        | 23  | 16.90%       |
| Nougat             | 24  | 11.40%       |
|                    | 25  | 7.80%        |
| Oreo               | 26  | 12.90%       |
|                    | 27  | 15.40%       |
| Pie                | 28  | 10.40%       |

Reasonable [Java 8 support](https://developer.android.com/studio/write/java8-support) was added in Api level 24 (Nougat). 

This leaves 42.1% of Android users which can't fully support java8 (even with desugar bytecode transformation).  I would say we definitely need to be able to produce java7 compatible code. 

Wherever immutables codegen (processor) itself should be java7 or java8 is debatable.

**Comment by Apreld01** [comment]
I believe this is a moot point @asereda-gs , google has stopped supporting Marshmallow since 2017 and the majority of the Play apps have been forced to update to 64-bit compatibility since then, requiring an update to their JDK.

**Comment by asereda-gs** [comment]
Why then rxjava still supports Java 6+ because of Android ?
This is applicable to v3 which is being released soon (currently RC4):

> continued support for Java 6+ & Android 2.3+

**Comment by Apreld01** [comment]
I don't think juxtaposing support for 1.6 and RxJava makes sense in this context. Jackson (arguably) a far larger (and widespread) library has dropped 1.6 support in 2017. 

https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.9

I think the main driver here is that the market for 1.7 is quite non existent and poses little to no benefit.

**Comment by asereda-gs** [comment]
Even your library example (Jackson v2.9) has java 7 support:

> No changes from 2.8: Java 7 features available for all modules 

I would like to hear a compelling reason to drop java 7 while major libraries (eg. rxjava and jackson) still support it. 

Also what exactly prevents you from using java8+ with immutables ? The only thing I can think of is java7 byte-code of `value-processor` module and that occurs only at compile-time. 

For example jackson v2.9 is also compiled with java7 target.

**Comment by Apreld01** [comment]
Guava, Kafka etc do not support java7, I think those are major libraries too, no? Ultimately as maintainers, you are free to do with what you wish, I believe that handicapping a system to an older version that sees diminishing support from major players is ultimately a detriment that hinders the platform.

**Comment by asereda-gs** [comment]
> Guava, Kafka etc do not support java7

Guava has a [separate artifact](https://guava.dev/) for Android (Java7) to this day ([v28.1](https://github.com/google/guava/releases/tag/v28.1) released Aug 2019)

> Guava provides two different “flavors”: one for use on a (Java 8+) JRE and one for use on Android or Java 7 or by any library that wants to be compatible with either of those. 

With kafka, **I think**, people are using HTTP proxies  (or the likes) and not connecting directly to brokers. 

> I believe that handicapping a system to an older version that sees diminishing support from major players is ultimately a detriment that hinders the platform.

I will be happy to consider your use-case. Let me know what functionality (which is not possible now) could have been achieved by dropping java7 support.

**Comment by Apreld01** [comment]
> I will be happy to consider your use-case. Let me know what functionality (which is not possible now) could have been achieved by dropping java7 support.

I did not address this earlier, apologize. For one, making changes to the codebase to take advantage of anything Java8+ is impossible. 

> Guava has a [separate artifact](https://guava.dev/) for Android (Java7) to this day ([v28.1](https://github.com/google/guava/releases/tag/v28.1) released Aug 2019)

Great! Is the suggestion here to fork and leave a version available for consumers of 1.7?

**Comment by asereda-gs** [comment]
> Great! Is the suggestion here to fork and leave a version available for consumers of 1.7?

No, the suggestion is that java 7 support is still important (for Android)

**Comment by asereda-gs** [comment]
>  For one, making changes to the codebase to take advantage of anything Java8+ is impossible.

Just to be clear, immutables doesn't impose java 7 on **your codebase**. Out of the box, one can  write java 8 POJOs (with `j.u.Optional`, `java.time.*`, `j.u.stream`  etc.) 

Immutables internals are still java7 (APT) and this, indeed, affects you if you want to contribute to the library.

**Comment by bjakke** [comment]
Hi,

My 2 cents on the issue. Additionally thanks for the great library/annotationprocessor :) (I think I have said that already at some point).

It really doesn't that much depend on what JVM versions are "supported" e.g. by oracle others, they still might be used (e.g. isolated factory networks). Additionally there are a lot of "non-standard" JVMs, e.g. some PLCs can run Java and you can run java inside Oracle DB. Additionally as far I'm aware some earlier versions of android are still used in developing countries, regardless of google's decisions on Play store (there are others). 

So in my opinion it is just a decision the maintainers of this library has to make (what to support), but I would hope that 

1) The minimum JVM version for generated code would be documented (a plus if you also document android API level), sorry if I have missed this.

2) Consider that updating the minimum version is somewhat major change (e.g. maybe in 3.0.0?). 

Android studio 3.0 supports some java 8 features (https://developer.android.com/studio/write/java8-support), but some require quite high android API version. I don't think those can be desugared, so it would need some care not to use those features for generated code (https://jakewharton.com/androids-java-8-support/). But also I do not have that much expertise in this.

So basically there is **not** a lot of need to update, unless the maintainers of this lib just cannot stand writing anonymous classes instead of lambdas ;)

P.S. Personally this change would not affect me (most projects are 11+, some are 8). Additionally unless you hit a bug, I don't see that much problems also for staying in old version of immutables.

P.S.2. Generally changing the minimum version could have great impact on libraries that might use immutables for generating their API (but then again at some point things will update past 9+ barrier, but then again old versions will still work as well). Sorry if this ended up long :)

**Comment by bmhm** [comment]
While I agreed in my second comment, I must revert that agreement.

Almost everyone else here is right. Higher java versions do support compiling java 7 sources to java 7 bytecode. And in this case this is a reasonable thing to do.

This java 7 bytecode can run in Java 8, 11, 14 applications on supported JVMs just fine. It's not as if the code compiled to java 7 by a Java 8 JDK is out of support. On the contrary!

That said, I agree to @bjakke -- the gain is non-existent, but the cost (a breaking change) is high.
